x/merkledb/proof.go

```diff
@@ -728,19 +752,20 @@ func valueOrHashMatches(value Maybe[[]byte], valueOrHash Maybe[[]byte]) bool {
 }
 
 // Adds each key/value pair in [proofPath] to [t].
-// For each proof node, adds the children that are < [start] or > [end].
-// If [start] is empty, no children are < [start].
-// If [end] is empty, no children are > [end].
+// For each proof node, adds the children that are
+// < [insertChildrenLessThan] or > [insertChildrenGreaterThan].
+// If [insertChildrenLessThan] is Nothing, no children are < [insertChildrenLessThan].
+// If [insertChildrenGreaterThan] is Nothing, no children are > [insertChildrenGreaterThan].
 // Assumes [t.lock] is held.
 func addPathInfo(
 	t *trieView,
 	proofPath []ProofNode,
-	startPath path,
-	endPath path,
+	insertChildrenLessThan Maybe[path],
+	insertChildrenGreaterThan Maybe[path],
```

Comment:
Now uses a `Maybe` to properly handle the case where we're verifying a proof that contains only the root. Namely, suppose that we had a range proof for range [`nil`, [0]]` where the only key-value pair is for the root.  In `RangeProof.Verify` we set `largestPath` to the empty path and called `addPathInfo` with `startPath` and `endPath` both the empty root. This causes `hasLowerBound` and `hasUpperBound` in `addPathInfo` to be `false`, and we did not add children greater than the root (the empty path), resulting in proof verification failing.

In other words, we need to use a `Maybe` to disambiguate between the case where the largest key in the proof is empty and the case where we shouldn't add any proof node children on the "right" side.